eduserver/input/input.go

```diff
@@ -90,6 +112,60 @@ func (t *EDUServerInputAPI) sendEvent(ite *api.InputTypingEvent) error {
 	return err
 }
 
+func (t *EDUServerInputAPI) sendToDeviceEvent(ise *api.InputSendToDeviceEvent) error {
+	devices := []string{}
+	localpart, domain, err := gomatrixserverlib.SplitID('@', ise.UserID)
+	if err != nil {
+		return err
+	}
+
+	if domain == t.ServerName && ise.DeviceID == "*" {
+		devs, err := t.DeviceDB.GetDevicesByLocalpart(context.TODO(), localpart)
+		if err != nil {
+			return err
+		}
+		for _, dev := range devs {
+			devices = append(devices, dev.ID)
+		}
+	} else {
+		devices = append(devices, ise.DeviceID)
```

Comment:
What happens in scenarios where the domain doesn't match and the device ID is `*`? Need commentary on why bundling `if domain == t.ServerName` makes sense at all.

Reply:
The idea is that, if it's targeted locally, we expand `*` into individual device names. If it isn't targeted locally then we just want to send the `*` unmodified, and the federation sender will (eventually, in a future PR) pick that up and shift it as-is.

Reply:
Please add exactly this commentary to the code :)